src/v/cluster/tm_stm.cc

```diff
@@ -0,0 +1,401 @@
+      [this, term, tx]() { return save_tx(term, tx); });
+}
+
+checked<tm_transaction, tm_stm::op_status>
+tm_stm::mark_tx_finished(kafka::transactional_id tx_id, tm_etag etag) {
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        return checked<tm_transaction, tm_stm::op_status>(
+          tm_stm::op_status::not_found);
+    }
+    if (ptx->second.etag != etag) {
+        return checked<tm_transaction, tm_stm::op_status>(
+          tm_stm::op_status::conflict);
+    }
+    ptx->second.status = tm_transaction::tx_status::finished;
+    ptx->second.etag = etag.inc_mem();
+    ptx->second.partitions = std::vector<tm_transaction::tx_partition>();
+    return checked<tm_transaction, tm_stm::op_status>(ptx->second);
+}
+
+checked<tm_transaction, tm_stm::op_status>
+tm_stm::mark_tx_ongoing(kafka::transactional_id tx_id, tm_etag etag) {
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        return checked<tm_transaction, tm_stm::op_status>(
+          tm_stm::op_status::not_found);
+    }
+    if (ptx->second.etag != etag) {
+        return checked<tm_transaction, tm_stm::op_status>(
+          tm_stm::op_status::conflict);
+    }
+    ptx->second.status = tm_transaction::tx_status::ongoing;
+    ptx->second.etag = etag.inc_mem();
+    ptx->second.tx_seq += 1;
+    ptx->second.partitions = std::vector<tm_transaction::tx_partition>();
+    return checked<tm_transaction, tm_stm::op_status>(ptx->second);
+}
+
+ss::future<tm_stm::op_status> tm_stm::re_register_producer(
+  kafka::transactional_id tx_id, tm_etag etag, model::producer_identity pid) {
+    auto timeout = config::shard_local_cfg().tm_sync_timeout_ms.value();
+    auto is_ready = co_await sync(timeout);
+    if (!is_ready) {
+        co_return tm_stm::op_status::unknown;
+    }
+
+    vlog(
+      clusterlog.trace,
+      "Registering existing tx: id={}, pid={}, etag={}",
+      tx_id,
+      pid,
+      etag);
+
+    auto term = _insync_term;
+    auto ptx = _tx_table.find(tx_id);
+    if (ptx == _tx_table.end()) {
+        co_return tm_stm::op_status::not_found;
+    }
+    if (ptx->second.etag != etag) {
+        co_return tm_stm::op_status::conflict;
+    }
+    auto tx = ptx->second;
+    tx.status = tm_transaction::tx_status::ongoing;
+    tx.pid = pid;
+    tx.tx_seq += 1;
+    tx.partitions = std::vector<tm_transaction::tx_partition>();
+
+    auto r = co_await _tx_locks.find(tx_id)->second->with(
+      [this, term, tx]() { return save_tx(term, tx); });
+
+    if (!r.has_value()) {
+        co_return tm_stm::op_status::unknown;
+    }
+    co_return tm_stm::op_status::success;
+}
+
+ss::future<tm_stm::op_status> tm_stm::register_new_producer(
+  kafka::transactional_id tx_id, model::producer_identity pid) {
+    auto timeout = config::shard_local_cfg().tm_sync_timeout_ms.value();
+    auto is_ready = co_await sync(timeout);
+    if (!is_ready) {
+        co_return tm_stm::op_status::unknown;
+    }
```

Comment:
I've made tm_sync_timeout_ms a property to prevent config reading on the hot path, but otherwise I it's too soon to compact the code, I might want to use different error codes for different methods when focus on returning more precise error codes to the client.